docs(marshal): expand smallcaps readability invariants (#2954)
endojs/endo: packages/marshal/docs/smallcaps-cheatsheet.md

| number           | Float64       | `Infinity`<br>`-Infinity`<br>`NaN`<br>`-0`<br>`7.1` | `"#Infinity"`<br>`"#-Infinity"`<br>`"#NaN"`<br>`"#-0"` // unimplemented<br>`7.1` |
| string           | String        | `'#foo'`<br>`'foo'`   | `"!#foo"` // special strings<br>`"foo"` // other strings |
| byteArray        | ByteArray     | `buf.toImmutable()`   | // undecided & unimplemented |
| passable symbols | Symbol        | `passableSymbolForName('foo')` | `"%foo"` // in transition |
| copyArray        | List          | `[a,b]`               | `[<a>,<b>]`          |

> | copyRecord       | Struct        | `{x:a,y:b}`           | `{<x>:<a>,<y>:<b>}`  |

Reviewer: Not in scope for this PR, but it would be nice to clarify that in this doc. I ended up finding the information in the ocapn spec, but besides that, most of the doc is pretty standalone.

> | copyRecord       | Struct        | `{x:a,y:b}`           | `{<x>:<a>,<y>:<b>}`  |

Author: Unresolving this one because it should be simple enough to do. I'll consider it as a drive-by.

> | copyRecord       | Struct        | `{x:a,y:b}`           | `{<x>:<a>,<y>:<b>}`  |

Author: Addressed by an additional bullet in the list below. Done.

> | copyRecord       | Struct        | `{x:a,y:b}`           | `{<x>:<a>,<y>:<b>}`  |

Reviewer 2: Suggestion: demonstrate key encoding and ordering by example.
```suggestion
| copyRecord       | Struct        | `{a}`<br>`{a,"#a":a,b}` | `{"a":<a>}` // only string keys<br>`{"!#a":<a>,"a":<a>,"b":<b>}` // keys sorted |
```

Author: Done

| error            | Error         | `TypeError(msg)`      | `{"#error":<msg>,"name":"TypeError"}` |

* The `-0` encoding is defined as above, but not yet implemented in JS.
* In JS, passable symbols are in transition from JavaScript symbols to their own representation
* The number after `"$"` or `"&"` is an index into a separate slots array.
* ***Special strings*** begin with any of the `!"#$%&'()*+,-` characters.
* `<expr>` is nested encoding of `expr`.
* To be passable, arrays, records, and errors must also be hardened.
* Structs [can only have string-named properties](https://github.com/endojs/endo/blob/master/packages/pass-style/doc/copyRecord-guarantees.md).
* Errors can also carry an optional `errorId` string property.
* We expect to expand the optional error properties over time.
* The ByteArray encoding is not yet designed or implemented.

## Readability Invariants

> For every JSON encoding with no special strings, the JSON and smallcaps decodings are the same.

Reviewer: I think copilot had a point here. What does it mean for a "JSON encoding" to be with a special string? Special string is in the realm of the value to be encoded (or that was decoded).

Author: @mhofman I took @gibson042 's suggestion below. Resolving.

> For every value with no special strings that round trips through JSON, the JSON and smallcaps encodings are the same.

Author: I don't understand. I'm not referring to any concrete encoding mandated by ocapn. Indeed, ocaps itself is likely to mandate something completely different. I only speaking here about the smallcaps vs JSON, and about the smallcaps concrete encoding of the ocapn abstract-syntax/data-model.

> For every value with no special strings that round trips through JSON, the JSON and smallcaps encodings are the same.

Author: Does this question make sense?

> In other words, for these simple values, ***you can ignore the differences between smallcaps and JSON***.

Reviewer: I still wish we defined smallcaps as a transformation of passable data into JS objects such that a plain `JSON.stringify` produces the string encoding described above, and that parsing that string with `JSON.parse` is guaranteed to produce a value that is "deep-equal" to the output of the smallcaps transformation.

| error            | Error         | `TypeError(msg)`      | `{"#error":<msg>,"name":"TypeError"}` |

* The `-0` encoding is defined as above, but not yet implemented in JS.
* In JS, passable symbols are in transition from JavaScript symbols to their own representation
* The number after `"$"` or `"&"` is an index into a separate slots array.
* ***Special strings*** begin with any of the `!"#$%&'()*+,-` characters.
* `<expr>` is nested encoding of `expr`.
* Since `x` and `y` can only be strings, `<x>` and `<y>` are the nested encodings of those strings.
* To be passable, arrays, records, and errors must also be hardened.
* Structs [can only have string-named properties](https://github.com/endojs/endo/blob/master/packages/pass-style/doc/copyRecord-guarantees.md).
* Errors can also carry an optional `errorId` string property.
* We expect to expand the optional error properties over time.
* The ByteArray encoding is not yet designed or implemented.

## Readability Invariants

For every JSON encoding with no special strings, the JSON and smallcaps decodings are the same.

> For every value with no special strings that round trips through JSON, the JSON and smallcaps encodings are the same.

Reviewer 2: I'm not sure how to interpret this claim. Does this rephrasing capture its intent?
```suggestion
If a value `v` round-trips through `JSON.parse(JSON.stringify(v))` and contains no special strings, then the smallcaps encoding of `v` is identical to `JSON.stringify(v)`.
```
(where "_round-trips_" is understood in a deep-equal sense and excludes -0 because `JSON.stringify(-0)` is "0", and "_contains no special strings_" includes top-level values, deep values, and both top-level and deep property names)
